### PR TITLE
Release a hold on a socket when closing it. (#46)

### PR DESCRIPTION
**Pull request by keshonok**
In some cases sock_put() was not called. Because of that, some
sockets were not closed properly: sk->sk_destruct() was not called.

### CODE REVIEW

File: tempesta_fw/sock.c [hunk 1]
```diff
@@ -440,6 +440,13 @@ ss_do_close(struct sock *sk)
 	}
 }
 
+static void
+ss_do_close(struct sock *sk)
+{
+	__ss_do_close(sk);
+	sock_put(sk);
+}
+
```
Comment (krizhanovsky) [inline tempesta_fw/sock.c]: Why can't we just call ss_close() which does sock_put()? Why do we need 3 functions for socket closing?

Comment (keshonok) [inline tempesta_fw/sock.c]: `ss_close()` is called where is can be used in process context, so in includes additional locking. `ss_do_close()` is called in SoftIRQ context, and it does not need any extra locking around it, the socket is most probably locked already anyway. `__ss_do_close()` is the main body of the close function, and it is therefore called from both `ss_close()` and `ss_do_close()`.

`sock_put()` should be called outside of locks, that's why it's not in `__ss_do_close()` (or `ss_do_close()` before this patch).

Comment (krizhanovsky) [inline tempesta_fw/sock.c]: Ok, makes sense. Just add appropriate comment to the functions.

**Comment by krizhanovsky** [comment]
Tested. Good to merge after applying mentioned comments.

**Pull request merged**